### ISSUE TITLE
Feat Req: Instance Normalization 

### ISSUE THREAD

**Issue by Arjun31415** [tracel-ai/burn#1249]
### Feature description

Implementation of Instance Normalization from the paper [Instance Normalization: The Missing Ingredient for Fast Stylization](https://arxiv.org/abs/1607.08022).

Pytorch Docs- 
1. [InstanceNorm1D](https://pytorch.org/docs/master/generated/torch.nn.InstanceNorm1d.html)
2. [InstanceNorm2D](https://pytorch.org/docs/master/generated/torch.nn.InstanceNorm2d.html)

### Feature motivation
Nice to have and a commonly used normalization layer.

**Comment by tushushu** [comment]
Hey there, may I work on this one?

**Comment by antimora** [comment]
@tushushu please go ahead!

**Comment by tushushu** [comment]
In pytorch, the inheritance levels are: Module -> _NormBase -> _InstanceNorm -> (InstanceNorm1d, InstanceNorm2d), let me try to figure out if we can do it this way in Burn.

**Comment by Arjun31415** [comment]
> In pytorch, the inheritance levels are: Module -> _NormBase -> _InstanceNorm -> (InstanceNorm1d, InstanceNorm2d), let me try to figure out if we can do it this way in Burn.

Inheritance does not exist in rust, and moreover not required. Check out the batch norm files in this repo to see how it is done without any base classes

**Comment by tushushu** [comment]
According to the [doc](https://www.tensorflow.org/addons/tutorials/layers_normalizations#instance_normalization_tutorial) of Tensorflow: Instance Normalization is special case of group normalization where the group size is the same size as the channel size (or the axis size).

Shall we implement the InstanceNormalization based on GroupNormalization?

@Arjun31415 @antimora

**Comment by tushushu** [comment]
In addition, the layer Normalization is also a special case of group normalization where the group size is 1. Does it make sense to reimplement LN based GN, too?

**Comment by antimora** [comment]
@louisfd and @nathanielsimard will be more knowledgable to suggest which route we should take.

**Comment by louisfd** [comment]
Hi @tushushu 
For InstanceNorm it seems right. Since GroupNorm is 
```rust
pub struct GroupNorm<B: Backend> {
    num_groups: usize,
    num_channels: usize,
    gamma: Option<Param<Tensor<B, 1>>>,
    beta: Option<Param<Tensor<B, 1>>>,
    epsilon: f64,
    affine: bool,
}
```
Then InstanceNorm could just be a group norm constructed with num_groups == num_channels. Would be interesting to offer it with the right name. 

For reimplementing LayerNorm, I think it could help adding cohesion instead of duplication, but first I'd like to know why LayerNorm's forward is 6 lines while GroupNorm is 88 lines... if using the same forward makes LayerNorm less efficient then not sure we should do it.

**Comment by tushushu** [comment]
> Hi @tushushu For InstanceNorm it seems right. Since GroupNorm is
> 
> ```rust
> pub struct GroupNorm<B: Backend> {
>     num_groups: usize,
>     num_channels: usize,
>     gamma: Option<Param<Tensor<B, 1>>>,
>     beta: Option<Param<Tensor<B, 1>>>,
>     epsilon: f64,
>     affine: bool,
> }
> ```
> 
> Then InstanceNorm could just be a group norm constructed with num_groups == num_channels. Would be interesting to offer it with the right name.
> 
> For reimplementing LayerNorm, I think it could help adding cohesion instead of duplication, but first I'd like to know why LayerNorm's forward is 6 lines while GroupNorm is 88 lines... if using the same forward makes LayerNorm less efficient then not sure we should do it.

Hey Louis,
After reading the codes of `GroupNorm` and `LayerNorm`, I found that:
1. `LayerNorm` does not check the input shape, but `GroupNorm` does.
2. `LayerNorm` has not implemented the `affine` feature.
3. `LayerNorm` uses `var_mean_bias` to calculate the mean and var.

Maybe let me tests the performance of `LayerNorm` and `GroupNorm` with group size=1 first.

**Comment by tushushu** [comment]
@louisfd @nathanielsimard  After testing the benchmark, the result of gn with n_group = 1 and ln is the same, but the gn's performance is better. The test and result are as below:

```rust
 #[test]
    fn compare_gn_ln() {
        const N: usize = 100;
        let arr = [(); N].map(|_| thread_rng().gen_range(0.0..1.0));
        let device = Default::default();
        let input = Tensor::from_data(Data::from([arr]), &device);

        let start = Instant::now();
        let module_ln = LayerNormConfig::new(N).init::<TestBackend>(&device);
        let output_ln = module_ln.forward(input.clone());
        let duration = start.elapsed();
        println!("Time elapsed in ln.forward is: {:?}", duration);

        let start = Instant::now();
        let module_gn = GroupNormConfig::new(1, N).init::<TestBackend>(&device);
        let output_gn = module_gn.forward(input.clone());
        let duration = start.elapsed();
        println!("Time elapsed in gn.forward is: {:?}", duration);

        // The output of the layer norm and group norm should be the same
        output_gn
            .to_data()
            .assert_approx_eq(&output_ln.to_data(), 3);
    }
```


```
    cargo test --release --package burn-core --lib -- nn::norm::layer::tests::compare_gn_ln --exact --nocapture

    const N: usize = 100;
    Time elapsed in ln.forward is: 364µs
    Time elapsed in gn.forward is: 49.75µs

    const N: usize = 1000;
    Time elapsed in ln.forward is: 424.542µs
    Time elapsed in gn.forward is: 61µs

    const N: usize = 10000;
    Time elapsed in ln.forward is: 588.875µs
    Time elapsed in gn.forward is: 202.458µs
```

**Comment by tushushu** [comment]
I've submitted a [PR](https://github.com/tracel-ai/burn/pull/1321) for you guys to review.

**Comment by laggui** [comment]
Completed in PR #1321
